packages/eslint-plugin/src/util/isArrayMethodCallWithPredicate.ts

```diff
@@ -0,0 +1,35 @@
+import { getConstrainedTypeAtLocation } from '@typescript-eslint/type-utils';
+import type {
+  ParserServicesWithTypeInformation,
+  TSESTree,
+} from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+
+const ARRAY_PREDICATE_FUNCTIONS = new Set([
+  'filter',
+  'find',
+  'findIndex',
+  'findLast',
+  'findLastIndex',
+  'some',
+  'every',
+]);
+
+export function isArrayMethodCallWithPredicate(
+  services: ParserServicesWithTypeInformation,
+  node: TSESTree.CallExpression,
+): boolean {
+  if (
+    node.callee.type === AST_NODE_TYPES.MemberExpression &&
+    node.callee.property.type === AST_NODE_TYPES.Identifier &&
+    ARRAY_PREDICATE_FUNCTIONS.has(node.callee.property.name)
+  ) {
+    const checker = services.program.getTypeChecker();
+    const type = getConstrainedTypeAtLocation(services, node.callee.object);
+    return tsutils
+      .typeParts(type)
+      .some(t => checker.isArrayType(t) || checker.isTupleType(t));
```

Comment:
[Bug] I think we should use `unionTypeParts` and `intersectionTypeParts` here, because the rule doesn't report on this code:

```ts
declare const array: (string[] & {foo: 'bar'}) | (number[] & {bar: 'foo'})
array.every(() => Promise.resolve(true));
``` 

[playground](https://deploy-preview-9955--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3rGUCeAXPABSYwCWKA5gNoC68AZPAN4BmSSRA5AEazcAvgEp4AHxIoArgFteIGPSat%2BMHhyRDhAKBz4AdCABuCvMWKiAvAD54ABRhIZ5NCH1w0SCCeIYYUkGFhAG4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)